File: src/main/java/com/nibado/projects/advent/y2016/Day02.java
package com.nibado.projects.advent.y2016;

import java.util.List;

import static com.nibado.projects.advent.Util.printAnswer;
import static com.nibado.projects.advent.Util.readResource;

public class Day02 implements Runnable {
    @Override
    public void run() {
        List<String> input = readResource("/2016/day02.txt");
        Numpad pad = new Numpad();
        String code = "";
        for(String i : input) {
            pad.apply(i);
            code += pad.get();
        }

        printAnswer(code);
    }

    public static void main(String... argv) {
        new Day02().run();
    }

    public static class Numpad {
        public int x;
        public int y;

        public static final char[][] numbers = {{'1', '2', '3'}, {'4', '5', '6'}, {'7', '8', '9'}};

        public Numpad() {
            this.x = 0;
            this.y = 0;
        }

        public void apply(String directions) {
            String[] parts = directions.split("");

            for(String p : parts) {
                Direction d = Direction.valueOf(p);
                apply(d);
            }
        }

        public void apply(Direction... directions) {
            for(Direction direction : directions) {
                x += direction.x;
                y += direction.y;

                if (x > 1) {
                    x = 1;
                }
                if (x < -1) {
                    x = -1;
                }
                if (y > 1) {
                    y = 1;
                }
                if (y < -1) {
                    y = -1;
                }
            }
        }

        public char get() {
            return numbers[y + 1][x + 1];
        }
    }

    public enum Direction {
        U(0, -1),
        R(1, 0),
        D(0, 1),
        L(-1, 0);

        public final int x;
        public final int y;

        Direction(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}

File: src/main/kotlin/com/nibado/projects/advent/y2016/Day02.kt
package com.nibado.projects.advent.y2016

import com.nibado.projects.advent.Day
import com.nibado.projects.advent.Direction
import com.nibado.projects.advent.Point
import com.nibado.projects.advent.resourceLines

object Day02 : Day {
    private val input = resourceLines(2016, 2).map { it.toCharArray().toList() }.toList()

    private val keypad1 = KeyPad(arrayOf(
            arrayOf('1', '2', '3'),
            arrayOf('4', '5', '6'),
            arrayOf('7', '8', '9')
    ))

    private val keypad2 = KeyPad(arrayOf(
            arrayOf(' ', ' ', '1', ' ', ' '),
            arrayOf(' ', '2', '3', '4', ' '),
            arrayOf('5', '6', '7', '8', '9'),
            arrayOf(' ', 'A', 'B', 'C', ' '),
            arrayOf(' ', ' ', 'D', ' ', ' ')
    ))

    override fun part1() = solve(keypad1)
    override fun part2() = solve(keypad2)

    private fun solve(keypad: KeyPad) : String {
        var solution = ""

        input.forEach {
            solution += keypad.apply(it)
        }

        return solution
    }

    private class KeyPad(val chars: Array<Array<Char>>) {
        private val width = chars.size
        private var pos = Point(width / 2, width / 2)

        fun apply(presses: List<Char>) : Char {
            presses.forEach {
                val dir = when(it) {
                    'U' -> Direction.NORTH
                    'L' -> Direction.WEST
                    'D' -> Direction.SOUTH
                    'R' -> Direction.EAST
                    else -> { throw IllegalArgumentException(it.toString())}
                }

                val newPos = pos.add(dir)
                if(newPos.inBound(width - 1, width - 1) && chars[newPos.y][newPos.x] != ' ') {
                    pos = newPos
                }
            }

            return chars[pos.y][pos.x]
        }
    }
}